### PR TITLE
rewrite rlpx; inline dpt and eth

### PR DESCRIPTION
**Pull request by fanatid**
work in progress..

### CODE REVIEW

File: lib/dpt/types.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+const address = {
+  encode: function (value) {
+    if (ip.isV4Format(value)) return ip.toBuffer(value)
+    if (ip.isV6Format(value)) return ip.toBuffer(value)
+    throw new Error(`Invalid address: ${value}`)
+  },
+  decode: function (buffer) {
+    if (buffer.length === 4) return ip.toString(buffer)
+    if (buffer.length === 16) return ip.toString(buffer)
+    throw new Error(`Invalid address buffer: ${buffer.toString('hex')}`)
```
Comment (fanatid) [inline lib/dpt/types.js]: Sometimes I receive address buffer with length 12, some ethereum client has bug?

File: lib/rlpx/peer.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+  this._socket.write(this._eciesSession.createBody(msg))
+  return true
+}
+
+Peer.prototype._sendHello = function () {
+  let data = [
+    util.int2buffer(BASE_PROTOCOL_VERSION),
+    this._clientId,
+    this._capabilities.map((obj) => [ new Buffer(obj.name), util.int2buffer(obj.version) ]),
+    util.int2buffer(this._port), // Q: always should be 16-bits ?
```
Comment (fanatid) [inline lib/rlpx/peer.js]: should it be encoded by `rlp`? or should be buffer with length 2? like in DPT

Comment (wanderer) [inline lib/rlpx/peer.js]: rlp I think. But there isn't a clear spec. So I you may just have to look at go-ethereum code to find out for sure.

File: lib/dpt/types.js (hunk 1 above)
Comment (wanderer) [inline lib/dpt/types.js]: the address maybe rlp encoded itself. rlp [x,x,x,x] have you tried that?

Comment (fanatid) [inline lib/dpt/types.js]: you mean that this buffer also can be encoded by rlp?

Comment (wanderer) [inline lib/dpt/types.js]: so after looking more closely at the code; it shouldn't be RLP encoded at this point. I'm not sure what can explain the lenght of 12

**Comment by fanatid** [comment]
I think PR is ready for review

**Comment by wanderer** [comment]
Whats the reason for the inlined DPT? Also I prefer to have the comments in the code then generate the docs from the comments

**Comment by wanderer** [comment]
Also I would like the [Steam interface back](https://github.com/ethereumjs/node-devp2p/blob/master/lib/stream.js). Each peer should just produce a steam of data. [For example](https://github.com/ethereumjs/node-blockchain-server/blob/master/index.js#L37) I think its ok for now to inline the packet parsing. But it the should just use the steam interface to parse the incoming data.  [For example](https://github.com/ethereumjs/node-blockchain-server/blob/master/index.js#L37)

**Comment by fanatid** [comment]
> What is the reason for the inlined DPT?

It's hard question, in common: how much modules we should create? DPT, RLPx, ETH will be enough? How deep we should modularize?

They uses common code (util.js), and for me was easier works when I seen DPT code in one project with RLPx/ETH.

I do not want to say that hold all stuff in one package is cool, but in some cases, why not?

> Also I prefer to have the comments in the code then generate the docs from the comments

JSDoc? You want that docs were generated from source code?

**Comment by wanderer** [comment]
> It's hard question, in common: how much modules we should create? DPT, RLPx, ETH will be enough? How deep we should modularize?

I have used the DPT several time by its self to generate stats about the network. So I think it make sense to modularize that now. As far as the subprotocols go, I'm ok inlining the packet parsing for now. If more subprotols start to utilize RLPx in the future will have to revisit then.

> JSDoc? You want that docs were generated from source code?

yeah it uses [Documention](https://github.com/ethereumjs/node-devp2p/blob/master/package.json#L11).

**Comment by fanatid** [comment]
message parsing in `node-blockchain-server` looks good, I should made same here

> Also I would like the Steam interface back. Each peer should just produce a steam of data. For example. I think its ok for now to inline the packet parsing. But it the should just use the steam interface to parse the incoming data. For example

I think inherits Stream interface here is not good, as I understand current solution will works only with 1 protocol at the one time.

> I have used the DPT several time by its self to generate stats about the network. So I think it make sense to modularize that now.

but you also can install this package and use only DPT..

> yeah it uses [Documention](https://github.com/ethereumjs/node-devp2p/blob/master/package.json#L11).

not a big fun of documentation generators in just (maybe that I seen not good modules)

**Comment by wanderer** [comment]
> as I understand current solution will works only with 1 protocol at the one time.

your are right! I never finished this part. But I wanted to generate the stream based the [capabilities](https://github.com/ethereumjs/node-blockchain-server/blob/master/index.js#L29-L31).

So for example 
```
  var network = new Network({
    address: '0.0.0.0',
    publicIp: '127.0.0.1',
    port: port,
    // peerDB: peerDB,
    secretKey: new Buffer('a153387bcc66f16b6aeaed404d3c3e2ec04f3a85c6942e9de107fb8b2f71e322', 'hex'),
    capabilities: {
      eth: 61,
      lol: 99,
    }
  })
```

Then when we got a new connection

```
network.on('connection', function (peer) {
       var lolstream = peer.lol.createStream()
      var lolstream = peer.eth.createStream()
})
```

Another good reason to use steams here is so that we have the nearlg the same API as node;s [net](https://nodejs.org/dist/latest-v6.x/docs/api/net.html#net_socket_remotefamily). Where each peers is like a socket. This shuold make it easier to use with other packages. 

> not a big fun of documentation generators in just (maybe that I seen not good modules)
 Show all checks

Yeah, there is no perfect solution but we have a lot of modules here.. ethereumjs-account, ethereumjs-util, ethereumjs-vm, merkle-triem ethereumjs-blockchain... ect ect, all use the document generation and its very nice that they are all uniform. But I'm open to suggestions

**Comment by fanatid** [comment]
> your are right! I never finished this part. But I wanted to generate the stream based the capabilities.

I already done this, but more big problem that `stream.Duplex` from node not determinate with data messages! Nobody guarantee that if you call `stream.write` twice, than what `_write` will be called in stream twice.. In my PR, if developer want own sub-protocol, he should just implement function `_handleMessage` and call `send` that he receive through constructor arguments.

> Yeah, there is no perfect solution but we have a lot of modules here.. ethereumjs-account, ethereumjs-util, ethereumjs-vm, merkle-triem ethereumjs-blockchain... ect ect, all use the document generation and its very nice that they are all uniform and in the same style. But I'm open to suggestions

Interface of this modules should be freezed, right? Writing documentation by hands is not pretty cool solution, but for JS right now I think it's better choice.. since [documentation](https://www.npmjs.com/package/documentation) already used is an amount of packages, I should return it back..

**Comment by wanderer** [comment]
this is now here https://github.com/fanatid/ethereumjs-devp2p

**Pull request closed without merging**